packages/modules/solaredge/test_meter_registers.py

```diff
@@ -0,0 +1,47 @@
+from unittest.mock import Mock
+
+import pytest
+
+from modules.common import modbus
+from modules.solaredge.config import (Solaredge, SolaredgeCounterConfiguration, SolaredgeCounterSetup,
+                                      SolaredgeExternalInverterConfiguration, SolaredgeExternalInverterSetup)
+from modules.solaredge.device import Device
+
+
+class Params:
+    def __init__(self,
+                 name: str,
+                 meter_id_counter: int,
+                 meter_id_inverter: int,
+                 synergy_units: int,
+                 expected_counter_power_register: int,
+                 expected_inverter_power_register: int):
+        self.name = name
+        self.meter_id_counter = meter_id_counter
+        self.meter_id_inverter = meter_id_inverter
+        self.synergy_units = synergy_units
+        self.expected_counter_power_register = expected_counter_power_register
+        self.expected_inverter_power_register = expected_inverter_power_register
+
+
+cases = [Params("standard", 1, 2, 0, 40206, 40380),
+         Params("inverter on meter id 3", 1, 3, 0, 40206, 40380),
+         Params("swap counter and inverter", 2, 1, 0, 40380, 40206),
+         Params("swap counter and inverter, counter on meter id 3", 3, 1, 0, 40380, 40206),
+         Params("synergy units 2", 1, 2, 2, 40256, 40430),
+         Params("synergy units 3", 1, 2, 3, 40276, 40450)
+         ]
+
+
+@pytest.mark.parametrize("params", cases, ids=[c.name for c in cases])
+def test_meter_registers_calculation(monkeypatch, params: Params):
+    synergy_mock = Mock(return_value=params.synergy_units)
+    monkeypatch.setattr(modbus.ModbusTcpClient_, "read_holding_registers", synergy_mock)
+    dev = Device(Solaredge())
+    dev.add_component(SolaredgeCounterSetup(
+        id=0, configuration=SolaredgeCounterConfiguration(meter_id=params.meter_id_counter)))
+    dev.add_component(SolaredgeExternalInverterSetup(
+        id=1, configuration=SolaredgeExternalInverterConfiguration(meter_id=params.meter_id_inverter)))
+
+    assert dev.components["component0"].registers.powers == params.expected_counter_power_register
+    assert dev.components["component1"].registers.powers == params.expected_inverter_power_register
```

Comment:
Du bist mit dem test auf unterschiedlichen Ebenen unterwegs. Das zeigt sich schon im Namen. Normalerweise haben wir eine Datei `something.py` und dazu eine Datei `something_test.py` die das testet, was in erster Datei steht. Selbst wenn ich das `test` in `test_meter_registers.py` nach hinten stelle gibt es keine `meter_registers.py`. Du könntest die Datei `meter_test.py` nennen, aber es würde nicht ganz stimmen, denn du testest einiges mit, was in `Device` liegt. `device_test.py` wäre besser, weil das die Ebene ist auf der du ansetzt. Aber dafür geht der Test ganz schön ins Detail was die Register betrifft.

Um das ganze besser testbar zu machen würde ich den Test splitten in die Aspekte die du testen willst. Ich sehe da zwei Aspekte:

1. Das was die Funktion `_update_regs` tut: Komponenten auf Basis der meter_id ein Register-Objekt zuweisen
2. Das was die Klasse `SolaredgeMeterRegisters` tut: Die richtigen Register berechnen.

Das ganze dann auch in zwei Dateien.

Für die Funktion `_update_regs` machst du dir das Leben leichter, wenn du die Abhängigkeit von `self` entfernst. Ggf. indem du eine Funktion außerhalb der Klasse daraus machst oder einen `@staticmethod`. Kann dann zum Beispiel so aussehen:
```python
def set_component_registers(components: Iterable[Any], synergy_units: int) -> None:
    # ...
```
Wenn du das so testest, brauchst du im Test die Klasse `Device` nicht instantieren, musst entsprechend auch den Modbus-Client nicht monkey-patchen und kannst die Funktion trotzdem richtig gut auf Herz und Nieren prüfen. Es reicht dann auch, wenn du prüfst, dass `meter_id` und `synergy_units` and `SolaredgeMeterRegisters` "durchgeschliffen" wird. Zum Beispiel indem du `SolaredgeMeterRegisters` monkeypatchst - in etwa so (pseudocode):
```python
def test_meter_registers_calculation(monkeypatch):
    # setup
    monkeypatch.setattr(modules.solaredge.device, "SolaredgeMeterRegisters", Mock(side_effect=lambda *args: args))
    counter = Mock(spec=SolaredgeCounter)
    external_inverter = Mock(spec=SolaredgeExternalInverter)
    components = ["ignore me", counter, external_inverter]

    # execution
    set_component_registers(components, synergy_units=1)

    # evaluation
    assert external_inverter.registers == (2, 1)
```
Das Mock von `SolaredgeMeterRegisters` gibt anstelle des instantierten Objekts die Argumentliste als tuple zurück. Da kannst du im Nachhinein prüfen ob ein Objekt in `registers` gespeichert wurde, welches aus einem Aufruf von `SolaredgeMeterRegisters` mit den richtigen Argumenten kam. Ohne dass du die ganze Logik, die sich in `SolaredgeMeterRegisters` befindet mittesten musst. Das kann dann in einen seperaten Test.